### PR TITLE
OCPBUGS-17810: temporarily remove cert observability fields, add storageversionmigration for machineconfigpools,controllerconfig

### PR DESCRIPTION
**Pull request by jkyros**
**- What I did**

- Removed certificate observability fields from the MachineConfigPool and ControllerConfig CRDs and types.go, because if they make it in for 4.14 ships, we won't be able to easily change them to the proper `metav1.Time`. 

- Made sure the MCO never sets the cert availability date fields to a value

- This also adds a `StorageVersionMigration` for `ControllerConfig` and `MachineConfigPool` so etcd can clean out anything it has stored for these types (see: https://github.com/openshift/kubernetes-kube-storage-version-migrator/blob/master/USER_GUIDE.md#check-if-migration-has-completed) 

- The certificate cleanup that was using these fields as part of the Image Registry Certificate work  https://github.com/openshift/machine-config-operator/blob/40dfd9b927452690effd3400fbc8f556a2b3ff92/pkg/controller/template/template_controller.go#L480is was always a no-op because it's not running in the daemon, so its functionality is unaffected -- it's deleting certificates from the controller pod that were never there.   

**- How to verify it**
 - Upgrade the MCO from a previous version to this PR, observe `ControllerConfig.status.controllerCertificates` and `MachineConfigPool.status.certExpirys` date fields are removed

**- Description for the changelog**
Temporarily remove certificate observability fields for API migration

Closes: [OCPBUGS-17810](https://issues.redhat.com//browse/OCPBUGS-17810)

### CODE REVIEW

**Comment by jkyros** [comment]
@deads2k I tried to do the "StorageVersionMigration" thing you described in our meeting today, but no matter what I do, it seems like my migrations are always "Failed" due to not being able to find the requested resources, e.g.
```
apiVersion: migration.k8s.io/v1alpha1
kind: StorageVersionMigration
metadata:
  annotations:
    include.release.openshift.io/ibm-cloud-managed: "true"
    include.release.openshift.io/self-managed-high-availability: "true"
    include.release.openshift.io/single-node-developer: "true"
  creationTimestamp: "2023-08-16T19:15:34Z"
  generation: 1
  name: machineconfiguration-controllerconfig-storage-version-migration
  ownerReferences:
  - apiVersion: config.openshift.io/v1
    controller: true
    kind: ClusterVersion
    name: version
    uid: 3de8bc74-e194-4482-a7e6-dd9171f427c1
  resourceVersion: "6565"
  uid: 0a1b4e9a-cd25-49a9-bcef-66055c8fc0ff
spec:
  resource:
    group: machineconfiguration.openshift.io
    resource: ControllerConfig
    version: v1
status:
  conditions:
  - lastUpdateTime: "2023-08-16T19:18:31Z"
    message: 'failed to list resources: the server could not find the requested resource'
    status: "True"
    type: Failed
```
I've tried all manner of casings and pluralization. Is this expected? (As in...if it doesn't have anything it needs to convert right now it will "fail" ?) 

The migrator logs are not particularly helpful: 
``` 
E0816 19:18:30.125873       1 kubemigrator.go:89] failed to list resources: the server could not find the requested resource
I0816 19:18:31.136219       1 kubemigrator.go:110] machineconfiguration-controllerconfig-storage-version-migration: migration running
E0816 19:18:31.144348       1 kubemigrator.go:121] failed to list resources: the server could not find the requested resource
E0816 19:18:31.144360       1 kubemigrator.go:130] machineconfiguration-controllerconfig-storage-version-migration: migration failed: failed to list resources: the server could not find the requested resource
```
The `flowcontrol` migrations that came with the cluster are showing the same results. 

To be clear, we weren't versioning our API here, this is v1 --> v1 (I don't know if that matters for how this thing works)

**Comment by jkyros** [comment]
Disregard, it looks like that worked during that upgrade test, which passed, I think we're probably alright here (I do have to rebase though because we merged something that changed the MCD test:
```
apiVersion: migration.k8s.io/v1alpha1
kind: StorageVersionMigration
metadata:
  annotations:
    include.release.openshift.io/ibm-cloud-managed: "true"
    include.release.openshift.io/self-managed-high-availability: "true"
    include.release.openshift.io/single-node-developer: "true"
  creationTimestamp: "2023-08-16T22:22:56Z"
  generation: 1
  name: machineconfiguration-machineconfigpool-storage-version-migration
  ownerReferences:
  - apiVersion: config.openshift.io/v1
    controller: true
    kind: ClusterVersion
    name: version
    uid: cae154fe-d210-49e1-8678-aa86ef9eb907
  resourceVersion: "84019"
  uid: c10294cc-6f6a-4b32-92b4-a8ca146e2065
spec:
  resource:
    group: machineconfiguration.openshift.io
    resource: machineconfigpools
    version: v1
status:
  conditions:
  - lastUpdateTime: "2023-08-16T22:22:57Z"
    status: "True"
    type: Succeeded
```

**Comment by jkyros** [comment]
Rebased, had to account for the changes to mcd_test.go (since we don't care about pause anymore)

File: install/0000_90_machine-config-operator_01_certificate_observability_storage_migraiton.yaml
Comment (djoshy) [inline install/0000_90_machine-config-operator_01_certificate_observability_storage_migraiton.yaml:1]: -typo in filename migraiton > migration

File: test/e2e/mcd_test.go [hunk 1]
```diff
@@ -964,10 +958,6 @@ func TestMCDRotatesCerts(t *testing.T) {
 	}); err != nil {
 		t.Errorf("Mismatch between on disk cert and in cluster cert: %v", err)
 	}
-
-	err = helpers.WaitForCertStatusToChange(t, cs, oldData)
-	require.Nil(t, err)
-
```
Comment (djoshy) [inline test/e2e/mcd_test.go:970]: Since we aren't pausing anymore, can we take out the pause comment at the function header?

**Review by djoshy — COMMENTED** [review]
LGTM, thanks for doing this, John. I put in a couple of nits that aren't a huge deal, they can be cleaned up later when we bring this back in. I'll let @cdoern give the final label.

File: install/0000_90_machine-config-operator_01_certificate_observability_storage_migraiton.yaml
Comment (jkyros) [reply install/0000_90_machine-config-operator_01_certificate_observability_storage_migraiton.yaml:1]: great catch, thank you !

**Comment by sergiordlr** [comment]
Verified upgrading from version
```
$ oc get clusterversion
NAME      VERSION                              AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.14.0-0.nightly-2023-08-11-055332   True        False         16m     Cluster version is 4.14.0-0.nightly-2023-08-11-055332
```



1. The ControllerConfig and the MCPs have the expiration dates:

```
$ oc get controllerconfig -o yaml

  controllerCertificates:
  - bundleFile: KubeAPIServerServingCAData
    notAfter: 2033-08-18 07:55:26 +0000 UTC
    notBefore: 2023-08-21 07:55:26 +0000 UTC
    signer: CN=admin-kubeconfig-signer,OU=openshift
    subject: CN=admin-kubeconfig-signer,OU=openshift
  - bundleFile: KubeAPIServerServingCAData

$ oc get mcp worker -o yaml 

status:
  certExpirys:
  - bundle: KubeAPIServerServingCAData
    expiry: 2033-08-18 07:55:26 +0000 UTC
    subject: CN=admin-kubeconfig-signer,OU=openshift
```



2. We upgrade the cluster to our current PR

```
    history:
    - acceptedRisks: |-
        Target release version="" image="registry.build03.ci.openshift.org/ci-ln-crl6m4b/release:latest" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified
        Forced through blocking failures: Multiple precondition checks failed:
        * Precondition "ClusterVersionRollback" failed because of "LowDesiredVersion": 4.14.0-0.ci.test-2023-08-21-075752-ci-ln-crl6m4b-latest is less than the current target 4.14.0-0.nightly-2023-08-11-055332, but rollbacks and downgrades are not recommended
        * Precondition "ClusterVersionRecommendedUpdate" failed because of "UnknownUpdate": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.14.0-0.nightly-2023-08-11-055332 to 4.14.0-0.ci.test-2023-08-21-075752-ci-ln-crl6m4b-latest is unknown.
      completionTime: "2023-08-21T09:54:16Z"
      image: registry.build03.ci.openshift.org/ci-ln-crl6m4b/release:latest
      startedTime: "2023-08-21T08:52:22Z"
      state: Completed
      verified: false
      version: 4.14.0-0.ci.test-2023-08-21-075752-ci-ln-crl6m4b-latest
    - completionTime: "2023-08-21T08:29:54Z"
      image: registry.ci.openshift.org/ocp/release@sha256:8a5507bf897252cab6d1957d9477bce45e7427f4f798450605d3503aed936594
      startedTime: "2023-08-21T08:08:36Z"
      state: Completed
      verified: false
      version: 4.14.0-0.nightly-2023-08-11-055332
```

3. After the upgrade the dates are removed from the MCP and controllerconfig resources

```
$oc get controllerconfig -o yaml
 
   controllerCertificates:
    - bundleFile: KubeAPIServerServingCAData
      signer: CN=admin-kubeconfig-signer,OU=openshift
      subject: CN=admin-kubeconfig-signer,OU=openshift
    - bundleFile: KubeAPIServerServingCAData
      signer: CN=kubelet-signer,OU=openshift
      subject: CN=kube-csr-signer_@1692605451

$ oc get mcp worker -o yaml 

  certExpirys:
  - bundle: KubeAPIServerServingCAData
    subject: CN=admin-kubeconfig-signer,OU=openshift
  - bundle: KubeAPIServerServingCAData
    subject: CN=kube-csr-signer_@1692605451
  - bundle: KubeAPIServerServingCAData

```

4. If we install this PR directly (no upgrade), the cluster is installed without problems and the controllerconfig and MCP  resources have no reference to certiricates' dates


We can add the qe-approved label

/label qe-approved

**Pull request merged**